nemo/collections/llm/recipes/gpt3_175b.py

```diff
@@ -0,0 +1,245 @@
+        virtual_pipeline_model_parallel_size=virtual_pipeline_parallelism,
+        context_parallel_size=context_parallelism,
+        sequence_parallel=sequence_parallelism,
+        gradient_as_bucket_view=True,
+        ckpt_async_save=True,
+        ckpt_parallel_load=True,
+        ddp=run.Config(
+            DistributedDataParallelConfig,
+            check_for_nan_in_grad=True,
+            grad_reduce_in_fp32=True,
+            overlap_grad_reduce=True,
+            overlap_param_gather=True,
+            average_in_collective=True,
+        ),
+    )
+
+    trainer = run.Config(
+        nl.Trainer,
+        accelerator="gpu",
+        accumulate_grad_batches=1,
+        callbacks=callbacks,
+        devices=num_gpus_per_node,
+        limit_test_batches=50,
+        limit_val_batches=32,
+        log_every_n_steps=10,
+        max_steps=max_steps,
+        num_nodes=num_nodes,
+        plugins=bf16_mixed(),
+        strategy=strategy,
+        use_distributed_sampler=False,
+        val_check_interval=2000,
+    )
+
+    return trainer
+
+
+@run.cli.factory(target=pretrain, name=NAME)
+def pretrain_recipe(
+    dir: Optional[str] = None, name: str = "default", num_nodes: int = 1, num_gpus_per_node: int = 8, fn=pretrain
+) -> run.Partial:
+    """
+    Create a pre-training recipe for GPT3 175B model.
+
+    This function sets up a complete configuration for pre-training, including
+    model, trainer, data, logging, optimization, and resumption settings.
+
+    Args:
+        dir (Optional[str]): Directory for saving logs and checkpoints.
+        name (str): Name of the pre-training run.
+        num_nodes (int): Number of compute nodes to use.
+        num_gpus_per_node (int): Number of GPUs per node.
+        fn (Callable): The pre-training function to use.
+
+    Returns:
+        run.Partial: Partial configuration for pre-training.
+
+    Examples:
+        CLI usage:
+            $ nemo llm pretrain --factory gpt3_175b
+            $ nemo llm pretrain --factory "gpt3_175b(num_nodes=64, name='my_175b_pretrain')"
+
+        Python API usage:
+            >>> recipe = pretrain_recipe(name="gpt3_175b_pretrain", num_nodes=64)
+            >>> print(recipe)
+
+    Note:
+        This recipe is optimized for the large 175B model and requires significant computational resources.
+    """
+    return run.Partial(
+        fn,
+        model=model(),
+        trainer=trainer(
+            num_nodes=num_nodes,
+            num_gpus_per_node=num_gpus_per_node,
+            callbacks=[run.Config(TimingCallback)],
+        ),
+        data=run.Config(MockDataModule, seq_length=2048, global_batch_size=2048, micro_batch_size=2),
+        log=default_log(dir=dir, name=name, tensorboard_logger=tensorboard_logger(name=name)),
+        optim=distributed_fused_adam_with_cosine_annealing(max_lr=0.9e-4),
+        resume=default_resume(),
+    )
+
+
+@run.cli.factory(target=pretrain, name=NAME + "_performance")
+def pretrain_recipe_performance(
```

Comment:
I would be a bit afraid that this will blow up the number of recipes that we show in the CLI. Since we allow the user to call the recipe with arguments, what about adding a new arg to the pretrain recipes that will dictate if we use the performance one or not?

Reply:
I agree @marcromeyn . But, should we do this as a follow-up PR as this would involve performance recipes for all models (I assume) and the interface for perf recipe/perf mode would also change. Reviewing would be difficult/erroneous?

Also, cc: @erhoo82 @JimmyZhang12

Reply:
@marcromeyn We need this performance recipe for every key model training because the baseline recipe is missing many performance features and it is hard to communicate with users without having a recipe that has all the important perf features. There is a high risk that a user can miss out many of them without asking us for the actual CLI command line every time.

Reply:
@ericharper As we discussed, the baseline recipe is for debugging and we need a performance recipe for all model training.

BTW, are we supposed to have different recipes for different sequence lengths? 
Also, one question on the global batch size here.
![image](https://github.com/user-attachments/assets/08f73989-4762-4dd3-b9ab-1d806b5c9e7b)

Reply:
Sequence lengths have their own recipe, for example: https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/llm/recipes/llama3_70b_64k.py

Reply:
@ericharper 
Yes, I see that it uses different a CP size for the target seqlen. So, I think it is only TP and CP mapping that needs to be changed. Do we need a full recipe only for this change? Also, I think the global batch size there is wrong.

Reply:
We decided previously to include the full recipe for this but we could revisit it. If there's a bug please send a fix.

@malay-nagda we can merge this as is, but please send a follow up PR with @marcromeyn 's suggestion